src/rules/sat_3sat.jl

```diff
@@ -0,0 +1,136 @@
+"""
+The reduction result of a general SAT problem to a 3-SAT problem.
+
+### Fields
+- `sat_source::Satisfiability{GT, T}`: the source general SAT problem.
+"""
+struct ReductionSATTo3SAT{T}
+    sat_source::Satisfiability{T}
+    sat_target::Satisfiability{T}
+    new_var_map::Dict{Symbol, Symbol}
+    inverse_new_var_map::Dict{Symbol, Symbol}
+end
+target_problem(res::ReductionSATTo3SAT) = res.sat_target
+
+function reduceto(::Type{<:Satisfiability}, sat_source::Satisfiability)
+    sat_source_renamed, new_var_map, inverse_new_var_map = rename_variables(sat_source)
+    sat_target = transform_to_3_literal_cnf(sat_source_renamed)
+    return ReductionSATTo3SAT(sat_source, sat_target, new_var_map, inverse_new_var_map )
+end
+
+function extract_solution(res::ReductionSATTo3SAT, sol)
+     
+    # inverse_map = res.inverse_new_var_map
+
+    # source_vars = res.sat_source
+    num_source_vars = num_variables(res.sat_source)
+    target_vars = variables( res.sat_target )
+    @assert length(sol[1]) == length(target_vars)
+    # num_target_vars = num_variables(res.sat_target)
+
+    all_original_solutions = Vector{Vector{Int64}}()
+
+    for sol_tmp in sol
+
+        original_solution = fill(-1, num_source_vars) 
+
+        for (i, new_var) in enumerate(target_vars)
+
+            new_var_str = string(new_var)
+            if startswith(new_var_str, "x_")
+                # original_var = inverse_map[new_var]
+                original_index = parse(Int, new_var_str[3:end])
+                original_solution[original_index] = sol_tmp[i]
+            end
+        end
+
+        push!(all_original_solutions, original_solution)
+    end
+
+    return unique( all_original_solutions )
+end
+
+# ----Useful functions----
+# 001: Function to rename variables in the CNF
+function rename_variables(sat::Satisfiability)
```

Comment:
Please check the `gensym` function:
https://docs.julialang.org/en/v1/base/base/#Base.gensym

Reply:
Maybe we can temporarily keep the "rename_variables" function.